emacs: teach elisp builders the finalAttrs pattern

jian-lin opened the pull request:
## Description of changes

~This PR causes **0 rebuilds**.~

[1]: https://github.com/NixOS/nixpkgs/pull/311459#issuecomment-2256597719

### Review advice

It is easier to review this PR commit by commit.

The first four commits are small fixes and cleaning.  Almost all work is done in the last commit.

### Performance overhead of eval time

|       package set        | before | after | changed |
|--------------------------|--------|-------|---------|
| emacs.pkgs.elpaPackages  |  1.925 | 1.935 | +0.35%  |
| emacs.pkgs.melpaPackages |  8.312 | 8.558 | +3.0%   |


The commands used here are

```
nix nixpkgs#hyperfine -- --warmup 2 --runs 10 'NIXPKGS_ALLOW_BROKEN=1 nix eval --include nixpkgs=$PWD --file . emacs.pkgs.melpaPackages --apply \'pkgSet: map (drv: drv.drvPath) (builtins.filter (p: p.type or null == "derivation") (builtins.attrValues pkgSet))\' --no-eval-cache >/dev/null'
```
```
nix nixpkgs#hyperfine -- --warmup 10 --runs 30 'NIXPKGS_ALLOW_BROKEN=1 nix eval --include nixpkgs=$PWD --file . emacs.pkgs.elpaPackages --apply \'pkgSet: map (drv: drv.drvPath) (builtins.filter (p: p.type or null == "derivation") (builtins.attrValues pkgSet))\' --no-eval-cache >/dev/null'
```

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jian-lin committed:
emacs: fix name when pname is overridden

Before

nix-repl> (emacs.pkgs.magit.overrideAttrs (old: { pname = old.pname + "-patched"; })).name
"emacs-magit-20240522.204"

After

nix-repl> (emacs.pkgs.magit.overrideAttrs (old: { pname = old.pname + "-patched"; })).name
"emacs-magit-patched-20240522.204"

jian-lin committed:
emacs: remove redundant doCheck from genericBuild

The default value of doCheck is false.

jian-lin committed:
emacs: handle propagatedBuildInputs correctly for genericBuild

Without this patch, if there is propagatedBuildInputs in the arguments
of genericBuild, it will override the value set by genericBuild.  With
this patch applied, the argument and the value set by genericBuild are
merged instead.

jian-lin committed:
emacs: stop generating autoload file for trivialBuild

When Emacs starts, package-activate-all finds autoload files and
loads them.  However, the autoload file generated by trivialBuild is
never picked up by package-activate-all.  In other words, this feature
never works.  So let's remove it.

jian-lin committed:
emacs: teach elisp builders the finalAttrs pattern

This commit causes 0 rebuilds.

The performance overhead of eval time is as follows:

|       package set        | before | after | changed |
|--------------------------|--------|-------|---------|
| emacs.pkgs.elpaPackages  |  1.925 | 1.935 | +0.35%  |
| emacs.pkgs.melpaPackages |  8.312 | 8.558 | +3.0%   |

The commands used here are

nix nixpkgs#hyperfine -- --warmup 2 --runs 10 'NIXPKGS_ALLOW_BROKEN=1 nix eval --include nixpkgs=$PWD --file . emacs.pkgs.melpaPackages --apply \'pkgSet: map (drv: drv.drvPath) (builtins.filter (p: p.type or null == "derivation") (builtins.attrValues pkgSet))\' --no-eval-cache >/dev/null'

nix nixpkgs#hyperfine -- --warmup 10 --runs 30 'NIXPKGS_ALLOW_BROKEN=1 nix eval --include nixpkgs=$PWD --file . emacs.pkgs.elpaPackages --apply \'pkgSet: map (drv: drv.drvPath) (builtins.filter (p: p.type or null == "derivation") (builtins.attrValues pkgSet))\' --no-eval-cache >/dev/null'

nixf-tidy-review commented on pkgs/applications/editors/emacs/build-support/elpa.nix:
```diff
@@ -3,21 +3,24 @@
+
 { pname
 , version
 , src
```
definition `src` is not used

jian-lin replied:
indirectly used here https://github.com/NixOS/nixpkgs/pull/330589/files#diff-73ae8b0d6ab3f94bf73f38648b4ae08138ce24ae49724306bd9aeb4daf00dbb6R36

To minimize diff, I do not change `args.src` to `src`.

nixf-tidy-review commented on pkgs/applications/editors/emacs/build-support/elpa.nix:
```diff
@@ -3,21 +3,24 @@
+libBuildHelper.extendMkDerivation { } genericBuild (finalAttrs:
+
 { pname
 , version
```
definition `version` is not used

nixf-tidy-review commented on pkgs/applications/editors/emacs/build-support/generic.nix:
```diff
@@ -21,21 +21,26 @@ let
 in
 
+libBuildHelper.adaptMkDerivation { } stdenv.mkDerivation (finalAttrs:
+
```
definition `pname` is not used

jian-lin replied:
indirectly used as `finalAttrs.pname` here https://github.com/NixOS/nixpkgs/pull/330589/files#diff-4d0cbd8553ca2ab2a1daac35346bf2ae8322d3d69cf5dc64f1413e84a784a13bR43

jian-lin replied:
Hey @inclyc, I cannot press "Resolve conversation" after fixing this.  It seems like a bug of your bot.

jian-lin replied:
It works now.  Weird.  Sorry for the noise.

nixf-tidy-review commented on pkgs/applications/editors/emacs/build-support/generic.nix:
```diff
@@ -21,21 +21,26 @@
 
+libBuildHelper.adaptMkDerivation { } stdenv.mkDerivation (finalAttrs:
+
 { pname
```
definition `version` is not used

jian-lin replied:
indirectly used as `finalAttrs.version` here https://github.com/NixOS/nixpkgs/pull/330589/files#diff-4d0cbd8553ca2ab2a1daac35346bf2ae8322d3d69cf5dc64f1413e84a784a13bR43

nixf-tidy-review commented on pkgs/applications/editors/emacs/build-support/trivial.nix:
```diff
@@ -2,9 +2,19 @@
-callPackage ./generic.nix envargs ({
+libBuildHelper.extendMkDerivation { } (callPackage ./generic.nix envargs) (finalAttrs:
+
+{ pname
```
definition `pname` is not used

nixf-tidy-review commented on pkgs/applications/editors/emacs/build-support/trivial.nix:
```diff
@@ -2,9 +2,19 @@
+
+{ pname
+, version
+, src
```
definition `src` is not used

nixf-tidy-review commented on pkgs/applications/editors/emacs/build-support/trivial.nix:
```diff
@@ -2,9 +2,19 @@
+libBuildHelper.extendMkDerivation { } (callPackage ./generic.nix envargs) (finalAttrs:
+
+{ pname
+, version
```
definition `version` is not used

nixf-tidy-review reviewed (commented):
❌ | Hi! [nixf-tidy](https://github.com/nix-community/nixd/tree/main/libnixf) code linter found some issues in your touched files.If you believe I'm reporting false positives, feel free to ping <code>@inclyc</code>

jian-lin commented:
Hey @Aleksanaa @inclyc, `nixf-tidy` reports false positives here.

Unused parameters are totally valid in the "build helpers" (or builders) use case, especially combined with the `finalAttrs` pattern.  These unused parameters server as the interface and documentation of the build helper, providing information about the expected parameters to both and the Nix interpreter and people reading the Nix expression:

```console
nix-repl> lib.functionArgs emacs.pkgs.melpaBuild
{ buildInputs = true; commit = true; ename = true; files = true; meta = true; nativeBuildInputs = true; packageRequires = true; pname = false; postInstall = true; postUnpack = true; preUnpack = true; propagatedBuildInputs = true; propagatedUserEnvPkgs = true; recipe = true; version = false; }
```

Currently, there are other PRs to teach build helpers the `finalAttrs` pattern, such as [go], [python] and [rust].  There is also a general [PR][general] to ease this transition.  Presumably there will be more of these false positives reported in the future.

[go]: https://github.com/NixOS/nixpkgs/pull/321791
[python]: https://github.com/NixOS/nixpkgs/pull/271387
[rust]: https://github.com/NixOS/nixpkgs/pull/194475
[general]: https://github.com/NixOS/nixpkgs/pull/234651

inclyc commented on pkgs/applications/editors/emacs/build-support/elpa.nix:
```diff
@@ -3,21 +3,24 @@
+libBuildHelper.extendMkDerivation { } genericBuild (finalAttrs:
+
 { pname
 , version
```
So if we actually remove this line, i.e. `version`, will this package misbehave?

jian-lin replied:
This is not a package.  It is a build helper (or builder).

It still works if `version` is removed.  However, as I said above, it should not be removed.

>  These unused parameters server as the interface and documentation of the build helper, providing information about the expected parameters to both and the Nix interpreter and people reading the Nix expression

inclyc commented on pkgs/applications/editors/emacs/build-support/melpa.nix:
```diff
@@ -62,17 +64,17 @@ in
-              ${lib.optionalString (files != null) ":files ${files}"})
+, recipe ? (writeText "${finalAttrs.pname}-recipe" ''
+    (${finalAttrs.ename} :fetcher git :url ""
+              ${lib.optionalString (finalAttrs.files != null) ":files ${finalAttrs.files}"})
```
there are no obvious way to infer `finalAttrs.*` comes from `files`, sadly.

jian-lin replied:
Yeah, this seems hard for a parser.  I guess some evaluation is need.  Maybe provide a way to disable the `nixf-tidy` check?

jian-lin commented:
My local run of `nixpkgs-review` tells me it is indeed **0 rebuilds** 🎉

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/we-added-a-linter-to-nixpkgs-checking-workflows/49722/16

jian-lin commented:
Actually, these variables are not "unused".  They are passed to `stdenv.mkDerivation` and are used there.  It is indeed a hard thing for a parser to understand. 😅

inclyc commented:
> Actually, those variables in `generic.nix` are not "unused". They are passed to `stdenv.mkDerivation` and are used there. 😅

I just kindly ask how they are passed to `stdenv.mkDerivation` ? Maybe we can explicitly use it in the attrset like `{ inherit files; }` ? 

> It is indeed a hard thing for a parser to understand.

The parser can only understand the variable being used directly by other variables but not do heavily value tracking about how the names are used. There are indeed some cases where function formals (a.k.a parameters) in nix language they are used in `builtins.functionArgs`, in such case we can ignore the warnings emitted.

> These unused parameters serve as the interface and documentation of the build helper, providing information about the expected parameters to both and the Nix interpreter and people reading the Nix expression

I don't think they do exactly provide "expected parameters" to *Nix interpreter* in this case, because we can remove it without having any side-effect. If they are the part of documentation why shouldn't they go under `/** */` as per [RFC 0415](https://github.com/NixOS/rfcs/pull/145)?

ShamrockLee commented on pkgs/applications/editors/emacs/build-support/melpa.nix:
```diff
@@ -62,17 +64,17 @@ in
-              ${lib.optionalString (files != null) ":files ${files}"})
+, recipe ? (writeText "${finalAttrs.pname}-recipe" ''
+    (${finalAttrs.ename} :fetcher git :url ""
+              ${lib.optionalString (finalAttrs.files or null != null) ":files ${finalAttrs.files}"})
```
One must not attempt to determine the existence of an attribute in `finalAttrs`, as this will lead to infinite recursion when evaluated.

The CI is right about the arguments `files` being unused. We need to pass `file` down the base build helper to ensure its existence, and then we can access `finalAttrs`.

jian-lin replied:
> One must not attempt to determine the existence of an attribute in `finalAttrs`, as this will lead to infinite recursion when evaluated.

Both packages with null `files` (`emacs.pkgs.git-undo`) and non-null `files` (`emacs.pkgs.urweb-mode`) build.  Also, ofborg evaluates without error.  So presumably there is no infinite recursion here.

ShamrockLee replied:
> Both packages with null `files` (`emacs.pkgs.git-undo`) and non-null `files` (`emacs.pkgs.urweb-mode`) build.

The error is not shown simply because the default value of `recipe` is not evaluated.

Try to pass the default value (e.g., under `passthru`) and access it, and you'll see the `infinite recursion` error.

jian-lin replied:
> The error is not shown simply because the default value of `recipe` is not evaluated.

I beg to differ.

The build process actually [uses][1] `recipe`.

Building a recipe also works.

```console
> cat (nix build --include nixpkgs=$PWD --file . emacs.pkgs.git-undo.recipe --print-out-paths)
(git-undo :fetcher git :url ""
          )
```

Here is a minimal working example to show this pattern works:

```console
nix-repl> f = self: {x = self.y or 666; }                                         

nix-repl> (lib.fix f).x
666

nix-repl> 
```

[1]: https://github.com/NixOS/nixpkgs/blob/0f44621c076ad3dc27225de13894e5c36e8989a7/pkgs/applications/editors/emacs/build-support/melpa.nix#L95-L99

ShamrockLee replied:
That's strange. I'll go ahead and check it out.

ShamrockLee replied:
You're right. The pattern that would cause infinite recursion is "specifying the value of an attribute based on its existence in `finalAttrs`."

ShamrockLee commented:
Other "truly unused" arguments are debatable. Should we keep them for documentation purposes or remove them to eliminate dead codes?

I agree with @inclyc that we should move them into the documentation. #308377 shows how such unused parts of set-pattern arguments could confuse.

inclyc commented:
Indeed yes because if declare the formal arg without actually using it (i.e. by referencing in `@arg` may results in a common mistake:

```
nix-repl> ({ a ? "foo" }@attrs: attrs.a) { }
error: attribute 'a' missing

       at «string»:1:23:

            1| ({ a ? "foo" }@attrs: attrs.a) { }
             |                       ^

nix-repl> 
```

In this case, while the lambda is called, `a` -> `"foo"`, however `attrs` -> `{ }`, and thus `attrs.a` is undefined.

So in our "builder" example, I suggest always explicitly USE the declaration like:

`{ foo, bar, ... }@args: someOtherFunction (args // { inherit foo bar })`

jian-lin commented:
> Indeed yes because if declaring a formal arg without actually using it (i.e. by referencing in `@arg`) may results in a common mistake:
> 
> ```
> nix-repl> ({ a ? "foo" }@attrs: attrs.a) { }
> error: attribute 'a' missing
> 
>        at «string»:1:23:
> 
>             1| ({ a ? "foo" }@attrs: attrs.a) { }
>              |                       ^
> 
> nix-repl> 
> ```
> 
> In this case, while the lambda is called, `a` -> `"foo"`, however `attrs` -> `{ }`, and thus `attrs.a` is undefined.
> 

I know this example a few days ago.

However, this example is not relevant here.  There is not any `args.a` or `attrs.a` in this PR.  It is `finalAttrs.a`.

> So in our "builder" example, I suggest always explicitly USE the declaration like:
> 
> `{ foo, bar, ... }@args: someOtherFunction (args // { inherit foo bar; })`

inclyc commented:
I mean why don't we need to `inherit` `files` here:

https://github.com/NixOS/nixpkgs/blob/7a34d0e63a1ef81af449c9d1cd7aaa583470aa5f/pkgs/applications/editors/emacs/build-support/melpa.nix#L82

like

```diff
- inherit packageBuild commit ename recipe; 
+ inherit packageBuild commit ename recipe files; 
```

Aleksanaa commented:
Then the unused error will move to there?

inclyc commented:
> Then the unused error will move to there?

I guess not, because `inherit` does not make new definition?

jian-lin commented:
Adding a unneeded `inherit file;` just to please the CI?  I do not think it is a good idea.

inclyc commented:
> Adding a unneeded `inherit file;` just to please the CI? I do not think it is a good idea.

Have you tried `.override`ing `files`  I just cannot understand why `finalAttrs` contains `files` attribute if it is not `inherited` like other names.

jian-lin commented:
> Have you tried `.override`ing `files`?

Parameters with added by `` is expected to be overridden with `overrideAttrs` (not `override`) likes this

```console
nix-repl> (emacs.pkgs.git-undo.overrideAttrs (old: { files = "foooooo"; })).files
"foooooo"
```

> I just cannot understand why `finalAttrs` contains `files` attribute if it is not `inherit`ed like other names.

Please have a look at `lib.adaptMkDerivation` and `lib.extendMkDerivation`.

jian-lin commented:
Sorry.  I somehow missed this.

> > Actually, those variables in `generic.nix` are not "unused". They are passed to `stdenv.mkDerivation` and are used there. 😅
> 
> I just kindly ask how they are passed to `stdenv.mkDerivation` ? 
> 

In `generic.nix`, `pname` and `version` is implicitly passed to `stdenv.mkDerivation` with the help of `lib.adaptMkDerivation`.  Note `// removeAttrs args handledArgs`.  Please have a look at `lib.adaptMkDerivation`.

> Maybe we can explicitly use it in the attrset like `{ inherit files; }` ?

Note that `files` is not in `generic.nix`.

> > It is indeed a hard thing for a parser to understand.
> 
> The parser can only understand the variable being used directly by other variables but not do heavily value tracking about how the names are used. There are indeed some cases where function formals (a.k.a parameters) in nix language they are used in `builtins.functionArgs`, in such case we can ignore the warnings emitted.
> 
> > These unused parameters serve as the interface and documentation of the build helper, providing information about the expected parameters to both and the Nix interpreter and people reading the Nix expression
> 
> I don't think they do exactly provide "expected parameters" to _Nix interpreter_ in this case, because we can remove it without having any side-effect.

By "providing information about the expected parameters to both the Nix interpreter", I mean you can check `lib.functionArgs emacs.pkgs.melpaBuild` in the repl.  I already gave an example there.

> If they are the part of documentation why shouldn't they go under `/** */` as per [RFC 0415](https://github.com/NixOS/rfcs/pull/145)?

Someone has to write the doc string.  Someone also has to make the doc system render these nix files.  I may do these things in the future, but not now.

inclyc commented:
> Please have a look at lib.adaptMkDerivation.

CC @ShamrockLee who authored those functions. WDYT?

ShamrockLee commented:
> > In `generic.nix`, `pname` and `version` is implicitly passed to `stdenv.mkDerivation` with the help of `lib.adaptMkDerivation`. Note `// removeAttrs args handledArgs`. Please have a look at `lib.adaptMkDerivation`.
> CC @ShamrockLee, who authored those functions. WDYT?

`lib.adaptMkDerivation` is designed *not* to pass arguments implicitly to the base build helper, in contrast to `lib.extendMkDerivations`. Instead, arguments are passed *explicitly* by `// removeAttrs args handledArgs`.

(People are already doing this a lot in existing build helper definitions, so we still need `lib.adaptMkDerivation` in addition to `lib.extendBuildHelper`.)

ShamrockLee commented:
IMHO, this issue is not specific to `lib.adaptMkDerivation` and `lib.extendMkDerivation`. People are already experiencing this in the current way of defining build helpers. An existing example can be found in the description of #308377.

The question is, should we keep syntactically unused arguments in the set pattern of a build helper? Should the set pattern and `__functionArgs` be considered a form of documentation?

CC @fricklerhandwerk, who devoted a lot of time to the documentation. What do you think about it?

jian-lin commented:
I add data about performance overhead [here](https://github.com/NixOS/nixpkgs/pull/330589#issue-2433776435).

jian-lin commented on pkgs/applications/editors/emacs/build-support/trivial.nix:
```diff
@@ -2,9 +2,19 @@
+}@args:
+
+{
   buildPhase = ''
```
I decide to ignore the user provided "previous" `args.buildPhase` because I consider `buildPhase` as the key component of this build helper.

That being said, users can still change `buildPhase` using `overrideAttrs` like this:

```nix
(trivialBuild {
  ...
}).overrideAttrs (finalAttrs: prevAttrs: {
  buildPhase = ...
})
```

The `buildPhase` and `installPhase` of other two build helpers, `elpaBuild` and `melpaBuild`, also follow this design.

 @ShamrockLee I see your related [comment][1] about factor lang.  Any thoughts here?

[1]: https://github.com/NixOS/nixpkgs/pull/287852#pullrequestreview-2208970569

ShamrockLee replied:
I'm unfamiliar with emacs packages, so commenting on the indispensability of the "key component" phases might be difficult.

A (counter) example for `buildGoModule` is that `apptainer` and `singularity` use the `make` and `make install` instead of `go build` and Golang-specific installation.

https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/virtualization/singularity/generic.nix#L275-L285

AndersonTorres replied:
> The user provided "previous" args.buildPhase is likely a mistake.

I am not sure if we should outsmart the user here.
You argue that there is an alternative overriding route for those really interested in shooting their own feet, but arguably this is more complicated.

Eval-time warnings are nicer, given that they can be turned off for those that swear to know what they are doing.

ShamrockLee replied:
Update:

*Nixpkgs Reference Manual* states that

> However, typically one only wants to add some commands to a phase, e.g. by defining postInstall or preFixup, as skipping some of the default actions may have unexpected consequences.

https://nixos.org/manual/nixpkgs/stable/#sec-stdenv-phases

IIUC, the statement means that one should typically use the pre- and post-attributes instead of overriding the whole phase. So, I would consider overriding the whole phase as a "use at your own risk" feature, even without additional warnings.

jian-lin replied:
> > Instead of ignoring `args.buildPhase`, maybe it is better to emit a warning at eval time
> 
> It is worth mentioning that such warnings are not suppressible (one can catch `throw`- based errors with `tryEval`, but not the warnings) and will appear when people `nix search` anything.

Presumably no elisp packages in Nixpkgs need their own `buildPhase` so such warnings are only seen by a small amount users with third party packages.

Note that a new argument can be easily introduced to control that warning.

> I am not sure if we should outsmart the user here.
You argue that there is an alternative overriding route for those really interested in shooting their own feet, but arguably this is more complicated.

👍

>  I would consider overriding the whole phase as a "use at your own risk" feature, even without additional warnings.

👍 I feel the same now.

jian-lin commented on pkgs/applications/editors/emacs/build-support/melpa.nix:
```diff
@@ -62,17 +64,17 @@ in
-genericBuild ({
+{
 
   elpa2nix = ./elpa2nix.el;
   melpa2nix = ./melpa2nix.el;
```
Should these two respect the value passed to `melpaBuild`?  I decide to ignore those values because I consider them as implementation detail of this build helper.

```nix
melpaBuild {
  elpa2nix = ./some-custom-file.el
  ...
}
```

The same question applies to `packageBuild`, `melpaVersion`

Any thoughts? @AndersonTorres 

related: https://github.com/NixOS/nixpkgs/pull/330589/files#r1697791893

fricklerhandwerk commented:
@ShamrockLee on the unused code: it's tricky... indeed leaving the attributes in the argument gives a hint at the interface, but it's highly idiomatic about the Nix language, you kinda have to know to know how to read that. I'd prefer to be explicit in the doc comments and not block this on improving automation, which will take a while even if you may decide it's a good thing to accommodate.

adisbladis commented:
> Since this PR has been open for 4 days, I am going to merge it as is in 3 days if no new discussions are made. If you want to review this but do not have time within three days, please let me know and I will wait for you.

I don't feel comfortable taking on the complexity of `extendDerivation` & friends vendored into Emacs lib.
I see https://github.com/NixOS/nixpkgs/pull/234651 as a blocker for this to be merged.

jian-lin commented:
The core of `extendDerivation` and `adaptMkDerivation` is quite small.

```nix
{
  extendMkDerivation' = mkDerivationBase: attrsOverlay:
    fpargs: (mkDerivationBase fpargs).overrideAttrs attrsOverlay;
  adaptMkDerivation' = mkDerivationBase: adaptArgs:
    fpargs: mkDerivationBase (finalAttrs: adaptArgs finalAttrs (lib.toFunction fpargs finalAttrs));
}
```

I would say `extendMkDerivation'` cannot be simplified further.  However, for `adaptMkDerivation'`, there are other possible implementations, such as doing the filtering inside `stdenv.mkDerivation` like the [rust PR][1] does.

Are you happy with vendoring those two simplified versions?  Or do you still think it is better to block this PR on https://github.com/NixOS/nixpkgs/pull/234651 to reach a consensus?

I do not know how to make progress on https://github.com/NixOS/nixpkgs/pull/234651.  That PR seems to have been blocked in the wait-for-review state for a long time.

Understanding the `finalAttrs` pattern for elisp builders does solve problems, such as making it easy to override some elisp [pakcages][2].

[1]: https://github.com/NixOS/nixpkgs/pull/194475/files#diff-282a02cc3871874f16401347d8fadc90d59d7ab11f6a99eaa5173c3867e1a160R300
[2]: https://github.com/NixOS/nixpkgs/pull/329709

jian-lin committed:
emacs: stop vendoring PR #234651

Previously, we vendor PR #234651 because we want to keep the old
behavior of filtering out packageRequires from the arguments we pass
to the underling stdenv.mkDerivation.  Doing so raises the concern
about the complexity of PR #234651.

Considering that passing packageRequires to stdenv.mkDerivation also
works well, we stop filtering it out and stop vendoring PR #234651.

Now, this PR only uses the existing interface of stdenv.mkDerivation.
Even though the name of the build helper is still extendMkDerivation',
it is nothing new and has been used in Nixpkgs, such as
php.buildComposerProject[1].

[1]: https://github.com/NixOS/nixpkgs/blob/f3834de3782b82bfc666abf664f946d0e7d1f116/pkgs/build-support/php/builders/v1/build-composer-project.nix#L108

jian-lin committed:
emacs: remove unused parameters to make nixf-tidy-review bot happy

jian-lin commented:
> I don't feel comfortable taking on the complexity of `extendDerivation` & friends vendored into Emacs lib.

I just pushed a new [commit] to stop vendoring https://github.com/NixOS/nixpkgs/pull/234651.  Its commit message is as follows:

> Previously, we vendor PR #234651 because we want to keep the old
behavior of filtering out packageRequires from the arguments we pass
to the underling stdenv.mkDerivation.  Doing so raises the concern
about the complexity of PR #234651.
>
> Considering that passing packageRequires to stdenv.mkDerivation also
works well, we stop filtering it out and stop vendoring PR #234651.
>
> Now, this PR only uses the existing interface of stdenv.mkDerivation.
Even though the name of the build helper is still extendMkDerivation',
it is nothing new and has been used in Nixpkgs, such as
php.buildComposerProject[1].

[1]: https://github.com/NixOS/nixpkgs/blob/f3834de3782b82bfc666abf664f946d0e7d1f116/pkgs/build-support/php/builders/v1/build-composer-project.nix#L108


@adisbladis I would like to if this commit resolves your concern about complexity. 

Unfortunately, since we pass `packageRequires` to `stdenv.mkDerivation` now, `packageRequires` is added to derivations of almost all elisp packages.  So this PR has to target the staging branch.

[commit]: https://github.com/NixOS/nixpkgs/pull/330589/commits/c8846d415fb23fc2f281b0c63b3b3b32203a8b2c

jian-lin committed:
emacs: make elpa2nix of elpaBuild consistent with melpaBuild

jian-lin committed:
emacs: inherit files in melpaBuild to make nixf-tidy CI happy

jian-lin commented:
Conflicts are resolved.

jian-lin commented on pkgs/applications/editors/emacs/build-support/generic.nix:
```diff
@@ -97,4 +97,4 @@ stdenv.mkDerivation (finalAttrs: ({
   '' + postInstall;
```
Almost all other attrs are overridden by user-defined values.  Applying that pattern to `postInstall` means changing `postInstall = '' .... '' + postInstall` to `postInstall = args.postInstall or '' ... ''`, which means if users define `postInstall`, their package never does native compilation.  I think this is very bad.  However, inconsistence is also not good.  WDYT?

AndersonTorres replied:
Well, well, well...

Searching some packages, I have found that e.g. `copyDesktopItems` injects itself at postInstallHooks. 

```nix
postInstallHooks+=(copyDesktopItems)
# rest of code
```

It implies that post installation hooks are "queued" but not destroyed/substituted.
(And it explains why we need to never forget the `runHook pre<phase-name>`...)

That being said, curiously the consistency is not in substituting pre/post phases, but in queueing them. 
So I vote for `+`

jian-lin replied:
Note that there are [two kinds][0] of preInstall hooks, one is defined in nix files with the attr `preInstall`, others are defined as an array called `preInstallHooks` in [setup hooks][1].  For the first kind of hook, it is indeed destroyed/substituted.  For the second kind of hook, generally it is queued using `preInstallHooks+=(...)`.

[0]: https://github.com/NixOS/nixpkgs/blob/c3aa7b8938b17aebd2deecf7be0636000d62a2b9/pkgs/stdenv/generic/setup.sh#L173
[1]: https://nixos.org/guides/nix-pills/20-basic-dependencies-and-hooks#setup-hooks

jian-lin replied:
Another question, which one of the two orders should be used for `+`, `postInstall = '' .... '' + postInstall` or `postInstall = postInstall + '' .... ''`?

jian-lin commented on pkgs/applications/editors/emacs/build-support/melpa.nix:
```diff
@@ -90,7 +94,7 @@ genericBuild ({
-    else version;
+    else finalAttrs.version);
 
   preUnpack = ''
```
Almost all other attrs are overridden by user-defined values.  Should we apply that pattern to `preUnpack` for consistency, i.e., changing `preUnpack = '' ... '' + preUnpack` to `preUnpack = args.preUnpack or '' ... ''`?  The same question applies to `postUnpack`.

AndersonTorres replied:
`preUnpack` suggests the idea we need to some preparation to the source so that it can be unpacked by the unpack phase of the builder we are dealing with. Further, this unpack phase is, at least conceptually, a black box.

Think in something like "oh the unpacker does not like dashes in the filename, let's change it to underscores first!" The `preUnpack` would change the filename, so that it can be passed to the black box.

So I would follow the same as I said to postInstall:  `+`  instead of `or`.

adisbladis commented on pkgs/applications/editors/emacs/build-support/melpa.nix:
```diff
@@ -62,24 +64,26 @@ in
-, recipe ? (writeText "${pname}-recipe" ''
-    (${ename} :fetcher git :url ""
-              ${lib.optionalString (files != null) ":files ${files}"})
+, recipe ? (writeText "${finalAttrs.pname}-recipe" ''
```
Not something to address here and now:
Do we really need to use `writeText` for the recipe? Couldn't it be done inline in the builder instead, skipping the intermediate derivation?

jian-lin replied:
Sounds good.  I created https://github.com/NixOS/nixpkgs/issues/334888 to track this potential improvement.
